packages/api/src/routes/medical/document.ts

```diff
@@ -108,29 +119,117 @@ router.post(
  * @return presigned url
  */
 router.get(
   "/downloadUrl",
+  asyncHandler(async (req: Request, res: Response) => {
+    const url = getDownloadUrl(req);
+    return res.status(MOVED_PERMANENTLY).json({ url });
+  })
+);
+
+/** ---------------------------------------------------------------------------
+ * GET /document/download-url
+ *
+ * Fetches the document from S3 and sends a presigned URL
+ *
+ * @param req.query.fileName The file name of the document in s3.
+ * @param req.query.conversionType The doc type to convert to.
+ * @return presigned url
+ */
+router.get(
+  "/download-url",
+  asyncHandler(async (req: Request, res: Response) => {
+    const url = getDownloadUrl(req);
+    return res.status(OK).json({ url });
+  })
+);
+
+/**
+ * POST /document/upload-url
+ *
+ * Uploads a medical document and creates a Document Reference for that file on the FHIR server.
+ *
+ * @param patientId - The ID of the patient.
+ * @body - The Partial<DocumentReference> draft to specify context for the file to be uploaded.
+ *
+ * @return documentReferenceID for the uploaded file.
+ * Refer to Metriport Documentation for more details:
+ * https://docs.metriport.com/medical-api/api-reference/document/post-upload-url
+ */
+router.post(
+  "/upload-url",
   asyncHandler(async (req: Request, res: Response) => {
     const cxId = getCxIdOrFail(req);
-    const fileName = getFromQueryOrFail("fileName", req);
-    const fileHasCxId = fileName.includes(cxId);
-    const type = getFrom("query").optional("conversionType", req);
-    const conversionType = type ? docConversionTypeSchema.parse(type) : undefined;
+    const patientId = getFromQueryOrFail("patientId", req);
+    const fileId = uuidv7();
+    const s3FileName = createS3FileName(cxId, patientId, fileId);
+    const docRefId = uuidv7();
+    const s3Key = s3FileName + "_upload_" + docRefId;
 
-    if (!fileHasCxId && !Config.isSandbox()) throw new ForbiddenError();
+    const organization = await getOrganization({ cxId });
+    if (!organization) throw new ForbiddenError(`Organization not found for CX ${cxId}`);
 
-    const url = await downloadDocument({ fileName, conversionType });
+    const docRefDraft = req.body;
+    docRefCheck(docRefDraft);
 
-    return res.status(OK).json({ url });
+    const docRef = pickDocRefParts(
+      docRefDraft,
+      organization,
+      docRefId,
+      patientId,
+      s3Key,
+      medicalDocumentsUploadBucketName
+    );
+    console.log("Updated DocumentReference:", JSON.stringify(docRef));
+
+    console.log(
+      "Trying to get the presigned url with bucket:",
+      medicalDocumentsUploadBucketName,
+      "Region",
+      region,
+      "and key:",
+      s3Key
+    );
+
+    const presignedUrl = await s3Utils.getPresignedUrl({
+      bucket: medicalDocumentsUploadBucketName,
+      key: s3Key,
+    });
+
+    // Make a temporary DocumentReference on the FHIR server.
+    await upsertDocumentToFHIRServer(cxId, docRef);
```

Comment:
ditto, thought 'draft' doc refs would be stored on the shared, temporary tenant on the FHIR server.

Or are you planning to update the places we search for doc refs of a customer to filter out the ones with `docStatus=preliminary`?

Reply:
We decided to keep these drafts on the cx tenant - [Context](https://metriport.slack.com/archives/C04DMKE9DME/p1698363856693339?thread_ts=1698361031.592839&cid=C04DMKE9DME)